crates/libcontainer/src/workload/mod.rs

```diff
@@ -46,6 +46,8 @@ pub trait CloneBoxExecutor {
 pub trait Executor: CloneBoxExecutor {
     /// Executes the workload
     fn exec(&self, spec: &Spec) -> Result<(), ExecutorError>;
+
+    fn validate(&self, spec: &Spec) -> Result<(), ExecutorError>;
```

Comment:
There are a few things to unpack here. I would group `before_exec` or similar functions into lifecycle hook category. OCI spec offers these. I am a little hesitant to expand this concept for the executor.

You raised a valid point that user would want to know the environment when validate runs. One potential solution is to extensively document the behavior. However, this will require the users to actually dig in the documentation (or comments) to understand this.

Currently, we are designing this based on a single usercase from runwasi. Is there another usecase we can look to?

Reply:
> Currently, we are designing this based on a single usercase from runwasi. Is there another usecase we can look to?

I don't know for now. Why not create a new error(maybe `ValidateError`) to narrow down usage for this interface?

Reply:
New error is hard. We provide an error interface for the trait without the actual implementation. There is no good way to know upfront what specific validation error should arise. Even with the executor error, we have a custom error as one of the type for this reason. The only error that can be handled is the `can't handle` error.

Reply:
I was a little short in my explanation. My point is this:
```rust
#[derive(Debug, thiserror::Error)]
pub enum ExecutorValidateError {
    #[error("{0} executor can't handle spec")]
    CantHandle(&'static str),
}

pub trait Executor: CloneBoxExecutor {
    /// Executes the workload
    fn exec(&self, spec: &Spec) -> Result<(), ExecutorError>;

    /// Validate if the spec can be executed by the executor. This step runs
    /// after the container init process is created, entered into the correct
    /// namespace and cgroups, and pivot_root into the rootfs. But this step
    /// runs before waiting for the container start signal.
    fn validate(&self, spec: &Spec) -> Result<(), ExecutorValidateError>;
}
```

I think `exec()` and `validate()` will return the different types of errors. They should be fundamentally separated.

Reply:
I see. Done.